common/src/main/java/com/gregtechceu/gtceu/api/gui/widget/LongInputWidget.java

```diff
@@ -0,0 +1,150 @@
+        super(x, y, width, height);
+
+        this.value = initialValue;
+        this.onChanged = onChanged;
+
+        buildUI();
+    }
+
+    private void buildUI() {
+        int buttonWidth = Mth.clamp(this.getSize().width / 5, 15, 40);
+        int textFieldWidth = this.getSize().width - (2 * buttonWidth) - 4;
+
+        this.addWidget(new ButtonWidget(0, 0, buttonWidth, 20,
+                new GuiTextureGroup(GuiTextures.VANILLA_BUTTON, getButtonTexture("-", buttonWidth)),
+                this::decrease
+        ).setHoverTooltips("gui.widget.incrementButton.default_tooltip"));
+
+
+        this.textField = new TextFieldWidget(buttonWidth + 2, 0, textFieldWidth, 20,
+                () -> String.valueOf(value),
+                stringValue -> this.setValue(Mth.clamp(Long.parseLong(stringValue), min, max))
+        );
+        this.updateTextFieldRange();
+        this.addWidget(this.textField);
+
+
+        this.addWidget(new ButtonWidget(buttonWidth + textFieldWidth + 4, 0, buttonWidth, 20,
+                new GuiTextureGroup(GuiTextures.VANILLA_BUTTON, getButtonTexture("+", buttonWidth)),
+                this::increase
+        ).setHoverTooltips("gui.widget.incrementButton.default_tooltip"));
+    }
+
+    private IGuiTexture getButtonTexture(String prefix, long buttonWidth) {
+        var texture = new TextTexture(prefix + "1");
+
+        if (!LDLib.isRemote()) {
+            return texture;
+        }
+
+        // Dynamic text is only necessary on the remote side:
+
+        long maxTextWidth = buttonWidth - 4;
+
+        texture.setSupplier(() -> {
+            long amount = GTUtil.isCtrlDown() ? GTUtil.isShiftDown() ? 512 : 64 : GTUtil.isShiftDown() ? 8 : 1;
+            String text = prefix + amount;
+
+            texture.scale(maxTextWidth / (float) Math.max(Minecraft.getInstance().font.width(text), maxTextWidth));
+
+            return text;
+        });
+
+        return texture;
+    }
+
+    private void increase(ClickData cd) {
+        this.changeValue(cd, 1);
+    }
+
+    private void decrease(ClickData cd) {
+        this.changeValue(cd, -1);
+    }
+
+    private void changeValue(ClickData cd, long multiplier) {
+        if (!cd.isRemote) {
+            long amount = cd.isCtrlClick ? cd.isShiftClick ? 512 : 64 : cd.isShiftClick ? 8 : 1;
+            this.setValue(Mth.clamp(value + (multiplier * amount), min, max));
+        }
+    }
+
+    public LongInputWidget setMin(long min) {
+        this.min = min;
+        updateTextFieldRange();
+
+        return this;
+    }
+
+    public LongInputWidget setMax(long max) {
+        this.max = max;
+        updateTextFieldRange();
+
+        return this;
+    }
+
+    public LongInputWidget setValue(long value) {
+        this.value = value;
+        onChanged.accept(value);
+
+        return this;
+    }
+
+    private void updateTextFieldRange() {
+        this.textField.setNumbersOnly(min, max);
+        this.setValue(Mth.clamp(this.value, min, max));
+    }
```

Comment:
merge it with your `IntInputWidget`? thry are similar
a `NumberInputWidget` is better.

Reply:
Hmm, been thinking about a more generic `NumberInputWidget` as well - will try to merge them together.